src/compiler/checker.ts

```diff
@@ -338,7 +338,15 @@ module ts {
                 // Locals of a source file are not in scope (because they get merged into the global symbol table)
                 if (location.locals && !isGlobalSourceFile(location)) {
                     if (result = getSymbol(location.locals, name, meaning)) {
-                        break loop;
+                        // Type parameters of a function are in scope in the entire function declaration, including the parameter
+                        // list and return type. However, local types are only in scope in the function body.
+                        if (!(meaning & SymbolFlags.Type) ||
+                            !(result.flags & (SymbolFlags.Type & ~SymbolFlags.TypeParameter)) ||
+                            !isFunctionLike(location) ||
+                            lastLocation === (<FunctionLikeDeclaration>location).body) {

```

Comment:
My second question has to do with this case
```ts
function f(x = a) {
    let a = 0;
}
```
`a` is not in scope where it is referenced in the parameter initializer. Yet it is in the locals of `f`. Presumably there is some logic to skip that symbol when we walk up, similar to what you've implemented here. My question is: Where is that logic, and can we combine your added logic with the logic for `let` and `const`?